drivers/adc/ad7616/ad7616.c

```diff
@@ -356,83 +358,218 @@ int32_t ad7616_set_mode(ad7616_dev *dev,
+		return ret;
+
+	axi_io_write(dev->core_baseaddr, AD7616_REG_UP_CTRL, AD7616_CTRL_RESETN);
+
+	return SUCCESS;
+}
+
+/**
+ * Initialize the AXI_AD7616 IP core device.
+ * @param device - The device structure.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t ad7616_core_setup(struct ad7616_dev *dev)
+{
+	uint32_t type;
+
+	axi_io_write(dev->core_baseaddr, AD7616_REG_UP_CTRL, 0x00);
+	mdelay(10);
+	axi_io_write(dev->core_baseaddr, AD7616_REG_UP_CTRL, AD7616_CTRL_RESETN);
+	axi_io_write(dev->core_baseaddr, AD7616_REG_UP_CONV_RATE, 100);
+	axi_io_write(dev->core_baseaddr, AD7616_REG_UP_CTRL,
+		     AD7616_CTRL_RESETN | AD7616_CTRL_CNVST_EN);
+	mdelay(10);
+	axi_io_write(dev->core_baseaddr, AD7616_REG_UP_CTRL, AD7616_CTRL_RESETN);
+	mdelay(10);
+
+	axi_io_read(dev->core_baseaddr, AD7616_REG_UP_IF_TYPE, &type);
+
+	if (type)
+		dev->interface = AD7616_PARALLEL;
+	else
+		dev->interface = AD7616_SERIAL;
+
+	return SUCCESS;
+}
+
 /**
  * Initialize the device.
  * @param device - The device structure.
  * @param init_param - The structure that contains the device initial
  * 					   parameters.
  * @return 0 in case of success, negative error code otherwise.
  */
-int32_t ad7616_setup(ad7616_dev **device,
-					 adc_core *core,
-					 ad7616_init_param init_param)
+int32_t ad7616_setup(struct ad7616_dev **device,
+		     struct ad7616_init_param *init_param)
 {
-	ad7616_dev *dev;
+	struct ad7616_dev *dev;
 	uint8_t i;
-	uint32_t if_type;
 	int32_t ret = 0;
 
-	dev = (ad7616_dev *)malloc(sizeof(*dev));
+	dev = (struct ad7616_dev *)malloc(sizeof(*dev));
 	if (!dev) {
 		return -1;
 	}
 
-	dev->core = core;
-	ad7616_core_read(*dev->core, AD7616_REG_UP_IF_TYPE, &if_type);
-	if (if_type)
-		dev->interface = AD7616_PARALLEL;
-	else
-		dev->interface = AD7616_SERIAL;
+	dev->core_baseaddr = init_param->core_baseaddr;
+	dev->offload_init_param = init_param->offload_init_param;
+	dev->reg_access_speed = init_param->reg_access_speed;
+	dev->dcache_invalidate_range = init_param->dcache_invalidate_range;
 
-	dev->spi_dev.chip_select = init_param.spi_chip_select;
-	dev->spi_dev.mode = init_param.spi_mode;
-	dev->spi_dev.device_id = init_param.spi_device_id;
-	dev->spi_dev.type = init_param.spi_type;
-	if (dev->interface == AD7616_SERIAL)
-		ret |= spi_init(&dev->spi_dev);
-
-	dev->gpio_dev.device_id = init_param.gpio_device_id;
-	dev->gpio_dev.type = init_param.gpio_type;
-	ret |= gpio_init(&dev->gpio_dev);
-
-	dev->gpio_reset = init_param.gpio_reset;
-	if (dev->gpio_reset >= 0) {
-		ret |= gpio_set_direction(&dev->gpio_dev,
-						dev->gpio_reset, GPIO_OUT);
-		ret |= gpio_set_value(&dev->gpio_dev,
-						dev->gpio_reset,
-						GPIO_HIGH);
+	ad7616_core_setup(dev);
+	if (dev->interface == AD7616_SERIAL) {
+		ret = spi_init(&dev->spi_desc, init_param->spi_param);
+		spi_engine_set_speed(dev->spi_desc, dev->reg_access_speed);
+	}
+	if (ret != SUCCESS)
+		return ret;
+
+	ret |= gpio_get_optional(&dev->gpio_hw_rngsel0,
```

Comment:
If you check the `ret`  value right after the function call is not necessary to `|` it

Reply:
you are right. will change.